File: backend/app/text_utils.py
from collections import Counter
import re

WORD_PATTERN = re.compile(r"[A-Za-z0-9']+")


def tokenize(text: str) -> list[str]:
    return [w.lower() for w in WORD_PATTERN.findall(text)]


def analyze_text(text: str) -> dict:
    words_original = WORD_PATTERN.findall(text)
    words = [w.lower() for w in words_original]
    word_count = len(words_original)
    character_count = len(text)
    line_count = len(text.splitlines()) if text else 0

    most_frequent_word = ""
    if words:
        counter = Counter(words)
        max_count = max(counter.values())
        most_frequent_word = sorted([w for w, c in counter.items() if c == max_count])[0]

    longest_word = max(words_original, key=len) if words_original else ""

    return {
        "word_count": word_count,
        "character_count": character_count,
        "line_count": line_count,
        "most_frequent_word": most_frequent_word,
        "longest_word": longest_word,
    }


def search_word(text: str, query: str) -> dict:
    query_normalized = query.strip().lower()
    if not query_normalized:
        return {"query": query, "count": 0, "positions": []}

    words = tokenize(text)
    positions = [idx for idx, w in enumerate(words, start=1) if w == query_normalized]

    return {
        "query": query,
        "count": len(positions),
        "positions": positions,
    }

File: frontend/src/textUtils.js
export function extractWordStats(text) {
  const words = (text.match(/[A-Za-z0-9']+/g) || []);
  const lower = words.map((w) => w.toLowerCase());

  const frequency = lower.reduce((acc, word) => {
    acc[word] = (acc[word] || 0) + 1;
    return acc;
  }, {});

  const mostFrequentWord = Object.keys(frequency)
    .sort((a, b) => {
      if (frequency[b] !== frequency[a]) return frequency[b] - frequency[a];
      return a.localeCompare(b);
    })[0] || '';

  const longestWord = words.sort((a, b) => b.length - a.length)[0] || '';

  return {
    wordCount: words.length,
    characterCount: text.length,
    lineCount: text ? text.split(/\r?\n/).length : 0,
    mostFrequentWord,
    longestWord
  };
}
